### DISCUSSION TITLE
more space used by object store in writeback mode

### DISCUSSION THREAD

**Discussion by shown1985** [juicedata/juicefs#3856] [General]
**What happened**:
使用--writeback参数挂载后，使用fio测试，占用的文件体积与实际不一致

**What you expected to happen**:
实际文件大小与占用空间保持一致

**How to reproduce it (as minimally and precisely as possible)**:
使用Format格式化文件系统
1. juicefs format --storage sftp --bucket 192.168.0.20:/share/CACHEDEV1_DATA/SFTP/ --access-key juicefs --secret-key 12345 --trash-days 0 redis://juicefs:12345@192.168.0.20:21379/4 sftp-test
2. 本地挂载 juicefs mount --writeback=true --background redis://juicefs:12345@192.168.0.20:21379/4 /mnt/test
3. 使用fio测试命令fio --rw=write --ioengine=sync --fdatasync=1 --file_service_type=sequential --directory=/mnt/test/ --size=100M --bs=4k --name=mytest
4. 使用juicefs info --raw /mnt/test/mytest.0.0查看
![image](https://github.com/juicedata/juicefs/assets/303880/9ace7c1a-fbe2-433a-9a9e-3927fda136fa)

发现有大量 size 4096的，导致体积变大。不带writeback参数的mount后，相同fio指令不存在类似现象。
正常情况如下：
![image](https://github.com/juicedata/juicefs/assets/303880/ff10edfa-b870-4092-902f-f2b21cf6c61d)

**Anything else we need to know?**

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version: juicefs version 1.0.4+2023-04-06.f1c475d
- Cloud provider or hardware configuration running JuiceFS: N/A
- OS (e.g `cat /etc/os-release`): CentOS Linux release 7.9.2009 (Core)
- Kernel (e.g. `uname -a`): Linux ASM-ML 3.10.0-1160.71.1.el7.x86_64 #1 SMP Tue Jun 28 15:37:28 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
- Object storage (cloud provider and region, or self maintained): SFTP
- Metadata engine info (version, cloud provider managed or self maintained): redis 7.0.4
- Network connectivity (JuiceFS to metadata engine, JuiceFS to object storage): LAN
- Others:

**Comment by davies** [comment]
fio 加上  --fdatasync 参数后会导致每次写入 4k 数据都会立即写入到 对象存储，导致大量碎片。写入过程中会触发碎片合并。在不使用 --writeback 时，碎片完全被合并了，符合预期。而 writeback 模式下，部分碎片未能合并，另外，可能还要检查对象存储中是否有未被清理的数据碎片。

我们会尝试看能否复现这个问题。

**Comment by shown1985** [comment]
我下午用webdav重新测试了这个问题，就像你说的，上传后碎片没有合并导致的，但是似乎webdav在writeback下上传结束后完成了碎片整理。可能因为使用的后端为sftp？晚些我再试下sftp

**Comment by shown1985** [comment]
> fio 加上 --fdatasync 参数后会导致每次写入 4k 数据都会立即写入到 对象存储，导致大量碎片。写入过程中会触发碎片合并。在不使用 --writeback 时，碎片完全被合并了，符合预期。而 writeback 模式下，部分碎片未能合并，另外，可能还要检查对象存储中是否有未被清理的数据碎片。
> 
> 我们会尝试看能否复现这个问题。

应该是这样的，当后端使用SFTP的时候，可以重现这个问题

**Comment by shown1985** [comment]
以下是juicefs客户端挂载日志

2023/05/24 22:53:16.033295 juicefs[20400] <INFO>: close session 2: %!s(<nil>) [base.go:369]
2023/05/24 22:55:03.230206 juicefs[20976] <INFO>: Meta address: redis://juicefs:****@192.168.0.20:21379/4 [interface.go:401]
2023/05/24 22:55:03.231702 juicefs[20976] <INFO>: Ping redis: 134.29µs [redis.go:2904]
2023/05/24 22:55:03.942372 juicefs[20976] <INFO>: Data use juicefs@192.168.0.20:/share/CACHEDEV1_DATA/SFTP/asm-sftp-test/ [mount.go:431]
2023/05/24 22:55:03.942706 juicefs[20976] <INFO>: Disk cache (/var/jfsCache/8a16eab1-bf55-493a-b141-9253163a7d69/): capacity (102400 MB), free ratio (10%), max pending pages (15) [disk_cache.go:94]
2023/05/24 22:55:03.970353 juicefs[21002] <INFO>: Meta address: redis://juicefs:****@192.168.0.20:21379/4 [interface.go:401]
2023/05/24 22:55:03.971818 juicefs[21002] <INFO>: Ping redis: 138.875µs [redis.go:2904]
2023/05/24 22:55:05.549276 juicefs[21002] <INFO>: Data use juicefs@192.168.0.20:/share/CACHEDEV1_DATA/SFTP/asm-sftp-test/ [mount.go:431]
2023/05/24 22:55:05.549447 juicefs[21002] <INFO>: Disk cache (/var/jfsCache/8a16eab1-bf55-493a-b141-9253163a7d69/): capacity (102400 MB), free ratio (10%), max pending pages (15) [disk_cache.go:94]
2023/05/24 22:55:05.551432 juicefs[21002] <INFO>: Create session 1 OK with version: 1.0.4+2023-04-06.f1c475d [base.go:289]
2023/05/24 22:55:05.552991 juicefs[21002] <INFO>: Prometheus metrics listening on 127.0.0.1:9567 [mount.go:161]
2023/05/24 22:55:05.553108 juicefs[21002] <INFO>: Mounting volume asm-sftp-test at /mnt/asm-k8s ... [mount_unix.go:181]
2023/05/24 22:55:58.357545 juicefs[21002] <WARNING>: Upload chunks/0/0/23_0_2048: ssh: handshake failed: read tcp 192.168.0.90:45490->192.168.0.20:22: read: connection reset by peer (try 1) [cached_store.go:398]
2023/05/24 22:55:58.411175 juicefs[21002] <WARNING>: Upload chunks/0/0/33_0_2048: ssh: handshake failed: read tcp 192.168.0.90:45494->192.168.0.20:22: read: connection reset by peer (try 1) [cached_store.go:398]
2023/05/24 22:55:58.665009 juicefs[21002] <WARNING>: Delete data blocks of slice 10 (2048 bytes): ssh: handshake failed: read tcp 192.168.0.90:45504->192.168.0.20:22: read: connection reset by peer [base.go:1008]
2023/05/24 22:55:58.668757 juicefs[21002] <WARNING>: Delete data blocks of slice 3 (2048 bytes): ssh: handshake failed: read tcp 192.168.0.90:45506->192.168.0.20:22: read: connection reset by peer [base.go:1008]
2023/05/24 22:55:58.668835 juicefs[21002] <WARNING>: Delete data blocks of slice 1 (2048 bytes): ssh: handshake failed: read tcp 192.168.0.90:45510->192.168.0.20:22: read: connection reset by peer [base.go:1008]
2023/05/24 22:55:58.672405 juicefs[21002] <WARNING>: Delete data blocks of slice 12 (2048 bytes): ssh: handshake failed: read tcp 192.168.0.90:45518->192.168.0.20:22: read: connection reset by peer [base.go:1008]
2023/05/24 22:56:08.522087 juicefs[21002] <INFO>: slow request: PUT chunks/0/0/6_0_2048 (<nil>, 10.200s) [cached_store.go:351]
2023/05/24 22:56:09.090222 juicefs[21002] <INFO>: slow request: DELETE chunks/0/0/14_0_2048 (<nil>, 10.418s) [cached_store.go:213]
2023/05/24 22:56:09.498731 juicefs[21002] <INFO>: slow request: DELETE chunks/0/0/8_0_2048 (<nil>, 10.837s) [cached_store.go:213]
2023/05/24 22:56:10.060212 juicefs[21002] <INFO>: slow request: DELETE chunks/0/0/9_0_2048 (<nil>, 11.399s) [cached_store.go:213]
2023/05/24 22:56:10.273596 juicefs[21002] <INFO>: slow request: PUT chunks/0/0/2_0_2048 (<nil>, 11.959s) [cached_store.go:351]
2023/05/24 22:56:10.477927 juicefs[21002] <INFO>: slow request: DELETE chunks/0/0/11_0_2048 (<nil>, 11.813s) [cached_store.go:213]
2023/05/24 22:56:11.014550 juicefs[21002] <INFO>: slow request: PUT chunks/0/0/19_0_2048 (<nil>, 12.669s) [cached_store.go:351]
2023/05/24 22:56:11.538258 juicefs[21002] <INFO>: slow request: PUT chunks/0/0/4_0_2048 (<nil>, 13.219s) [cached_store.go:351]
2023/05/24 22:56:11.628685 juicefs[21002] <INFO>: slow request: DELETE chunks/0/0/5_0_2048 (<nil>, 12.967s) [cached_store.go:213]
2023/05/24 22:56:11.777593 juicefs[21002] <INFO>: slow request: DELETE chunks/0/0/2_0_2048 (<nil>, 13.116s) [cached_store.go:213]
2023/05/24 22:56:12.010910 juicefs[21002] <INFO>: slow request: PUT chunks/0/0/24_0_2048 (<nil>, 13.652s) [cached_store.go:351]

**Comment by SandyXSD** [comment]
看起来是某些请求失败了导致碎片合并/清理没有完全完成，可以尝试手动执行 `juicefs gc --compact --delete` 来清理。
